src/ResponseBuilder.cpp

```diff
@@ -0,0 +1,176 @@
+#include "ResponseBuilder.hpp"
+
+/**
+ * @brief Construct a new ResponseBuilder object
+ *
+ * @param configFile Configuration file.
+ * @param fileSystemPolicy File system policy. Can be mocked if needed.
+ */
+ResponseBuilder::ResponseBuilder(const ConfigFile& configFile, const FileSystemPolicy& fileSystemPolicy)
+	: m_configFile(configFile)
+	, m_activeServer(configFile.serverConfigs.begin())
+	, m_fileSystemPolicy(fileSystemPolicy)
+	, m_isFirstTime(true)
+{
+	initMIMETypes();
+}
+
+/**
+ * @brief Set the active server.
+ *
+ * The active server is used for location searching.
+ * @param activeServer Server to be set as active.
+ */
+void ResponseBuilder::setActiveServer(const std::vector<ServerConfig>::const_iterator& activeServer)
+{
+	m_activeServer = activeServer;
+}
+
+/**
+ * @brief Get the built response.
+ *
+ * The stream is returned as c_str() since it is reused.
+ * If str() would be returned, old parts of the stream would be included.
```

Comment:
Not quite: the return value is constructed from the const char* ptr which points to the raw string buffer in the stringstream. 

> Doesn't this imply unnecessary conversions (from and to std::string)?

Stringstream uses a string buffer, which can be accessed via .str(). Since its a string we have access to its function c_str().

In contrast to classic c strings, std::string doesn't need a zero terminator. It tracks the size of the string itself. This means a std::string can have "Hello\0World" (a string with a zero terminator inbetween) and correctly return 11 as size, where strlen would return 5.
.c_str() returns (in general) a pointer to the raw buffer of the string, with a null terminator guaranteed at the end. To be more efficient this generally means (since C11 is standard I think) that a std::string buffer has a zero terminator at the end, > you don't have to allocate for a c_str() call. 
c_str() also has to have constant complexity, so regardless of string size it should need the same time. Therefore its more or less guaranteed (even pre C11) that you get a pointer to the underlying buffer with a zero terminator, since malloc and copy would not be constant.

Even though the std::string itself doesn't need a zero terminator, the buffer oftentimes is allocated +1 > mainly for compatibility reasons for a lot of C functions.

This means by passing stringstream.str() or stringstream.str().c_str() you can control which constructor of std::string gets called with no extra cost.

> Why not just clear the stringstream?

If we "clear" the stringstream by setting it to empty string we have two performance implications:
1. we need to copy the empty string
2. we don't know what happens with the allocated buffer

ad 1. When we set the stringstream to empty string we need to construct a string object and copy it into the buffer. I don't exactly know how it works but at least in godbolt it shows more code output for `ss.str("");` instead of `ss.seekp(std::ios::beg);`. Interestingly if we just copy an empty string - `ss.str(std::string))` - the difference is smaller.

See here for godbolt compiler explorer: [LINK](https://godbolt.org/z/PjdqPzx6M)

ad 2. If we never clear the buffer and just reuse the previously used thing we dont need to reallocate. I'm not sure if the buffer gets shrinked to size when you copy an empty buffer into it. But if we never clear it it can't happen. This means that if the ResponseBuilder creates a big message, the buffer stays big.

With seekp we just return to the beginning and add new stuff. Since most stuff gets added as if it was a c string, it also sets a new zero terminator.  Example:
```
stringstream ss;
ss << "Hello World";
ss.seekp(std::ios::beg)
ss << "Short";
// ss.str() is now "Short\0World"
```
When we would now use the std::string constructor and the use for example send() we still would only see "Short" since send() expects a zero terminated string, and it is terminated after "Short". But: we would also copy the never used stuff. I we use the const char* we only copy upto the zero temrinator.

> Why not clear it after sending?

We can get the same message multiple times as long as we don't build a new one. I don't see added value in clearing it gets used once. Granted: we won't use it multiple times, but why limit functionality with no added benefit?

Reply:
Oida. How much time did you spend to figure out this

Reply:
I read the article I posted. Played around in Godbolt to confirm.

Reply:
Which article? There is only the link to Godbolt.

Reply:
This one: http://cplusplus.bordoon.com/speeding_up_string_conversions.html 

Posted in Discord regarding reusing stringstream.
The article mentions using seekp and return as c_str().
It does not spell everything out, which is why I checked with godbolt > for example it doesn't say anything about "disadvantage" of setting it to "".